# tuberculosis-polymorph-llm-model-pipeline
 This project is a bioinformatics tool & python based pipeline for Modeling the Association of Gene Polymorphisms and Drug Resistance in Mycobacterium tuberculosis Using Deep Learning and Large Language Models

# Tools
Tools/library used in the pipeline :
- enaWebTools : https://github.com/enasequence/enaBrowserTools
- fastqc : https://www.bioinformatics.babraham.ac.uk/projects/fastqc/
- trimmomatic : https://github.com/timflutre/trimmomatic
- cutadapt : https://cutadapt.readthedocs.io/en/stable/guide.html#basic-usage
- blast-py : https://biopython.org/docs/dev/Tutorial/chapter_blast.html
- kraken-tools : https://github.com/jenniferlu717/KrakenTools
- bowtie2 : https://purduercac-applications.readthedocs.io/en/latest/Biocontainers/bowtie2/bowtie2.html
- samtools/pysam : https://www.htslib.org/ or https://pysam.readthedocs.io/en/stable/
- annovar : https://annovar.openbioinformatics.org/en/latest/user-guide/startup/
- tb-profiler : https://github.com/jodyphelan/TBProfiler
- scikit-learn : https://github.com/scikit-learn/scikit-learn
- tensorflow : https://pypi.org/project/tensorflow/
- pytorch : https://pypi.org/project/torch/
- Other library needed in python3 is Numpy,Pandas,sklearn,matplotlib, etc.

# Flowchart
<p align="center">
<img src="/img/llm-mtb-polymorp-pipeline.drawio.png" width="480" title="pipeline">
</p>

# What to do next 
- Preprocess
- SNP & Feature extraction
- LLM & DL Modelling
